crates/sortition/src/sortition.rs

```diff
@@ -488,30 +758,48 @@ impl Handler<CiphernodeRemoved> for Sortition {
 }
 
 impl Handler<GetNodeIndex> for Sortition {
-    type Result = Option<u64>;
+    type Result = ResponseFuture<Option<(u64, Option<u64>)>>;
 
-    /// Return the index of `address` in the size-`size` committee for `seed`
-    /// on `chain_id`. If the chain has not been initialized, returns `None`.
-    ///
-    /// Errors while accessing persisted state or parsing the address are
-    /// reported on the event bus and surfaced here as `None`.
-    #[instrument(name = "sortition_contains", skip_all)]
     fn handle(&mut self, msg: GetNodeIndex, _ctx: &mut Self::Context) -> Self::Result {
-        self.list
-            .try_with(|map| {
+        let node_state_manager = self.node_state_manager.clone();
+        let bus = self.bus.clone();
+
+        // Get the sortition backends synchronously
+        let backends_snapshot = self.list.get();
+
+        Box::pin(async move {
+            // Query NodeStateManager for fresh state
+            let node_state_snapshot = if let Some(manager) = node_state_manager {
+                manager.send(GetNodeState).await.ok().flatten()
```

Comment:
This seems really heavy handed to me. We just want an index and to do that we appear to get the entire NodeState global HashMap from the NodeStateManager that contains the state of every single node on every single chain and pass it to sortition. Actors are meant to only share data by message passing and we want to limit the size of the messsages we pass. To my eyes we should consider ditching `GetNodeState` in favour of a redesign.